common/web/types/src/consts/virtual-key-constants.ts

```diff
@@ -161,16 +161,42 @@ export const ISOVirtualKeyMap: KeyMap = [
   [ k.K_SPACE ],
 ];
 
+export const JISVirtualKeyMap: KeyMap = [
+  // [Hankaku/Zenkaku] 1 2 3 4 5 6 7 8 9 0 - ^ ¥ [bksp]
+  [ k.K_1, k.K_2, k.K_3, k.K_4, k.K_5, k.K_6, k.K_7, k.K_8, k.K_9, k.K_0, k.K_HYPHEN, k.K_EQUAL, k.K_oE2 /* YEN? */ ],
```

Comment:
Not 1000.0‰ sure about that k_oE2 at the end there.

Reply:
@miloush maybe?  The windows docs said that the ABNT2 key was also used for RO / backslash.

Reply:
OK so the OADG standard says that OADG109 and OADG109A keyboards have the extra key next to backspace and next to right shift (OADG104 has neither).

The standard specifies the scan codes as follows:

| key | Set 1 make | Set 1 break | Set 2 make | Set 2 break | Set 3 make | Set 3 break |
|--|--|--|--|--|--|--|
|next to backspace|7D| FD| 6A| F0 6A |5D|F0 5D|
|next to right shift|73|F3|51|F0 51|51|F0 51|

The standards also shows AX and J-3100 keyboards but I take the liberty to assume you are not interested in those (again, just saying JIS is very vague).

I do have an OADG109A keyboard and can confirm that the scan codes match the standard, set 1 on Windows. The key caps in the standard are as follows:

| key | OADG109 | OADG109A | my hardware |
|--|--|--|--|
| next to backspace | ![image](https://user-images.githubusercontent.com/10546952/228505463-f53c2645-05a4-44ab-9dd4-3bce9027e1e0.png) | ![image](https://user-images.githubusercontent.com/10546952/228505559-fc7b9c18-dfc5-4bd3-a588-cb60a138acda.png) |  ![image](https://user-images.githubusercontent.com/10546952/228506444-35d2db56-c549-4121-8b6c-192277c53f96.png) |
| next to right shift | ![image](https://user-images.githubusercontent.com/10546952/228506562-51e8415b-96e8-43f6-bb21-554f5b763b16.png) | ![image](https://user-images.githubusercontent.com/10546952/228506670-d11d0ead-9ca4-432d-a5bd-422fd8ba7153.png) | ![image](https://user-images.githubusercontent.com/10546952/228506802-a33807da-05c0-45ac-9c7a-1d9dbb1fa784.png) |

The standard does not formally define the key outputs other than in these figures.

Now the meaty part. As you probably know, the SC to VK mapping on Windows is part of the driver dll and as such is theoretically arbitrary, although in practice Windows ships _mostly_ one mapping. Unfortunately _mostly_ applies especially to the east Asian keyboards. You can change whether you have 101/106/109 etc. key in the settings. That changes which VKs are generated.

For Japanese, there are two settings which result in two different mappings:
| setting | next to backspace | next to shift |
|--|--|--|
| English keyboard (101/102 key) | invalid (0xFF) | VK_ABNT_C1 (0xC1) |
| Japanese keyboard (106/109 key) | VK_OEM_5 (0xDC) | VK_OEM_102 (0xE2) |

I hope this helps, good luck.

Reply:
This is by the way the reason why once you are on EA hardware, such as Japanese laptop in this case, you won't be able to type common characters that are on these VK when using any of the western layouts. I am hoping that LDML will be able to fix that.

Reply:
As for keys right to L, you get VK_OEM_PLUS, VK_OEM_1, VK_OEM_6 for 106/109 setting and VK_OEM_1, VK_OEM_7, VK_OEM_5 for 101/102 setting. The latter corresponds to colon, quote and backslash on US layout indeed. I don't have anyone to ask at the moment but I would assume that if you have one of these keyboards, you would use the 106/109 setting.

Reply:
Thank you @miloush! About to fly to Thailand, will be back in 10 days, so I probably won't have much useful to say until then :grin:, but your feedback here is much appreciated. I have always struggled to correlate my reality to the various standards out there!

> I hope this helps, good luck.

🤣 Thank you 🍀

Reply:
FWIW, on Windows, Keyman translates to scan codes and then back to a 'standard' kbdus-based layout (with tweak for abnt2 extra key). I don't remember exactly what we do with Japanese keyboards, would need to test to be sure.

Reply:
@miloush and all, i'm going to merge this but filed https://github.com/keymanapp/keyman/issues/8552 to fine tune the ABNT2 and JIS scancodes.   Per the discussion over on Keyboard-SC, we should be using metadata from CLDR once available.  This PR has other structural and build improvements, so I'd like to move it forward…thanks…